### ISSUE TITLE
Create an alternative for tuple calls before OTP21 release

### ISSUE THREAD

**Issue by amatalai** [appunite/mockery#22]

**Comment by teamon** [comment]
@amatalai Can you add some more context?

**Comment by amatalai** [comment]
@teamon ofc, yes

Due to this PR  https://github.com/erlang/otp/pull/1499 it will be necessary to add something like `@compile :tuple_calls` in order to use tuple calls. At the moment Mockery is using tuple calls to pass some data about mocks to function which handles most of logic and it means that you would need to add this compile attribute to every module where you want to mock something. It seems unacceptable to me.

I have an idea to replace tuple calls by generating small module with `:"$handle_undefined_function"` that will mimic tuple call behavior. In fact I already have working solution locally, but I'm not completely satisfied with it yet.

**Comment by teamon** [comment]
That's one of my early implementations https://gist.github.com/teamon/bad0891c02c26ed774890e04bb998e15#file-twin-ex-L9 (still running in one project, waiting to be replaced by mockery ;])

**Comment by amatalai** [comment]
I know it. Your implementation is where I learned about `$handle_undefined_function` 😈

**Comment by teamon** [comment]
Btw, Dialyzer is complaining about the tuple syntax

```erlang
The call {'Elixir.Mockery.Proxy', 'Elixir.MyDep', 'nil'}:'new'...) requires that {'Elixir.Mockery.Proxy', 'Elixir.MyDep', 'nil'} is of type atom() not {'Elixir.Mockery.Proxy', 'Elixir.MyDep', 'nil'}
```

@amatalai any chance to publish your work in progress to github branch? ;)

**Comment by amatalai** [comment]
It's known that dialyzer is complaining about tuple calls. 

My wip solution needs some cleanup, but I'll try to publish it in next few days.

**Comment by amatalai** [comment]
released Mockery 2.2 to hex
